### PR TITLE
[Reviewer: Krista] Add UTs for matching functions

### PR DESCRIPTION
**Pull request by eleanor-merry**
UTs for functions in https://github.com/Metaswitch/cpp-common/pull/628

### CODE REVIEW

File: src/ut/utils_test.cpp [hunk 1]
```diff
@@ -172,3 +172,103 @@ TEST_F(UtilsTest, RemoveBracketsFromBareIPv6Address)
...
+}
+
+TEST_F(UtilsTest, BracesWildcardMatch)
+{
+  EXPECT_TRUE(uris_user_match("tel:![4]{4}!", "tel:4444"));
+}
+
+TEST_F(UtilsTest, BracesWildcardMatchExtra)
+{
+  EXPECT_TRUE(uris_user_match("tel:![4]{4}!", "tel:44444"));
```
Comment (Krista-Rogers) [inline src/ut/utils_test.cpp]: Should this pass?

**Review by Krista-Rogers — APPROVED** [review]

File: src/ut/utils_test.cpp [hunk 2]
```diff
@@ -172,3 +172,103 @@ TEST_F(UtilsTest, RemoveBracketsFromBareIPv6Address)
 {
   EXPECT_EQ("::1", remove_brackets_from_ip("::1"));
 }
+
+TEST_F(UtilsTest, IsAWildcard)
+{
+  EXPECT_TRUE(is_wildcard_uri("sip:!.!*!@domain"));
```
Comment (graemerobertson) [inline src/ut/utils_test.cpp]: These are so cheap to write that I think it would be worth adding a few more `EXPECT_TRUE` statements to this test, specifically testing some of the following things...
* user part doesn't start with `!`
* user part doesn't finish with `!`
* More `!`
* Include other regex characters

I'm not sure how valuable those last two things are; it feels like they might be valuable if we changed the implementation of `is_wildcard_uri` ever.

File: src/ut/utils_test.cpp [hunk 3]
```diff
@@ -172,3 +172,103 @@ TEST_F(UtilsTest, RemoveBracketsFromBareIPv6Address)
...
+}
+
+TEST_F(UtilsTest, IsNotAWildcard)
+{
+  EXPECT_FALSE(is_wildcard_uri("sip:!.*@domain"));
```
Comment (graemerobertson) [inline src/ut/utils_test.cpp]: Similar to above, I think we should have more tests here...
* Longer URIs with `!` in different places
* No `!`
* tel URIs?
* `!` in the domain part (is that allowed?)

File: src/ut/utils_test.cpp (hunk 2 above)
Comment (graemerobertson) [inline src/ut/utils_test.cpp]: These are so cheap to write that I think it would be worth adding a few more `EXPECT_TRUE` statements to this test, specifically testing some of the following things...
* user part doesn't start with `!`
* user part doesn't finish with `!`
* More `!`
* Include other regex characters

File: src/ut/utils_test.cpp (hunk 3 above)
Comment (graemerobertson) [reply src/ut/utils_test.cpp]: Can params contain an `@` and/or a `!`?

File: src/ut/utils_test.cpp (hunk 1 above)
Comment (graemerobertson) [reply src/ut/utils_test.cpp]: +1

**Review by graemerobertson — APPROVED** [review]
Looks ok...

I'd quite like to test these functions by somehow generating random regexes and random strings to match against them, because it feels like the space of possible regexes and possible strings is enormous and it's difficult to test all the edge cases (or even define where the edges are). Unfortunately I can't think of a good way to do that. Also, I'm not so worried because the actual regex matching is just done by some other library which is probably/presumably/hopefully fine...

File: src/ut/utils_test.cpp (hunk 1 above)
Comment (eleanor-merry) [reply src/ut/utils_test.cpp]: I think it should - it only shouldn't pass if the regex was `^[4]{4}$`

Comment (Krista-Rogers) [reply src/ut/utils_test.cpp]: ok cool
